### PR TITLE
Add and use expect methods to hir.

### PR DESCRIPTION
**Pull request by WaffleLapkin**
[The future has come](https://github.com/rust-lang/rust/pull/106090/files#r1070062462).

r? @Nilstrieb

tbh I'm not even sure if it's worth it

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -3005,6 +3047,127 @@ impl Item<'_> {
     pub fn item_id(&self) -> ItemId {
         ItemId { owner_id: self.owner_id }
     }
+
+    /// Expect an [`ItemKind::ExternCrate`] or panic.
+    #[track_caller]
+    pub fn expect_extern_crate(&self) -> Option<Symbol> {
+        let ItemKind::ExternCrate(s) = self.kind else { unreachable!() };
```
Comment (compiler-errors) [inline compiler/rustc_hir/src/hir.rs:3054]: these should probably be `bug!`s

Comment (compiler-errors) [reply compiler/rustc_hir/src/hir.rs:3054]: even more ideally, they'd say "expected an extern crate, got `{:?}`" or something -- makes it much easier at a glance (or a stack trace submitted on an issue) to know what's up

Comment (WaffleLapkin) [reply compiler/rustc_hir/src/hir.rs:3054]: I don't think `bug!`s are available at this point in the crate graph :(

Comment (compiler-errors) [reply compiler/rustc_hir/src/hir.rs:3054]: nooooooo

ok well can u at least improve the panic message?

Comment (WaffleLapkin) [reply compiler/rustc_hir/src/hir.rs:3054]: Yes, I'll improve the messages (not today tho, I'm tired and tired)

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -3590,6 +3768,185 @@ impl<'hir> Node<'hir> {
...
+    #[track_caller]
+    pub fn expect_infer(self) -> &'hir InferArg {
+        let Node::Infer(this) = self else { self.expect_failed("an infer") };
+        this
+    }
+
+    #[track_caller]
+    fn expect_failed(&self, expected: &'static str) -> ! {
+        panic!("expected {expected} node, found {self:?}")
+    }
```
Comment (Nilstrieb) [inline compiler/rustc_hir/src/hir.rs:3949]: This file is way too big and should be split apart into like defs, helpers, expects etc. at some point :(
But it's fine for this PR

**Review by Noratrieb — COMMENTED** [review]
Looks good, looking forward to follow up PRs using it in more places :)

File: compiler/rustc_hir/src/hir.rs (hunk 2 above)
Comment (WaffleLapkin) [reply compiler/rustc_hir/src/hir.rs:3949]: yeah, a lot of files are like this :(

Comment (Nilstrieb) [reply compiler/rustc_hir/src/hir.rs:3949]: :(

**Pull request merged**